Forward source requests to debugpy as fallback

vidartf opened the pull request:
~If the source cannot be looked up by the `path` argument, forward it to debugpy as it might be able to resolve it using the sourceReference field.~

EDIT: If the `sourceReference` is non-zero, forward it to debugpy, as the specification requires the reference to be used in those cases.

Proposed fix to #870.

TODO: Write some tests.

vidartf committed:
Forward source requests to debugpy as fallback

If the source cannot be looked up by the `path` argument, forward it to debugpy as it might be able to resolve it using the sourceReference field.

JohanMabille commented on ipykernel/debugger.py:
```diff
@@ -442,9 +442,7 @@ async def source(self, message):
-            reply['success'] = False
-            reply['message'] = 'source unavailable'
-            reply['body'] = {}
+            reply = await self._forward_message(message)
```
See my [comment](https://github.com/ipython/ipykernel/issues/870#issuecomment-1049008052); I think the right implementation should be to handle the request in the kernel if and only if the attribute `sourceReference` is 0. Notice that the `source` attribute is optional, accoriding to the [DAP](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Source), therefore this implementation is not totally compliant.

vidartf replied:
I've had another go at this, trying to follow the spec more closely this time.

vidartf committed:
Align debugger source handler with DAP spec

Give `sourceReference` priority. Not immediately clear which location should be given priority if soruceReference and source.sourceReference are both defined and different values. Spec says that they should be the same, so we can claim that this is not our problem.

r3m0t commented on ipykernel/debugger.py:
```diff
@@ -434,17 +434,23 @@ async def source(self, message):
-                }
+        source_ref = message["arguments"].get("sourceReference", 0)
+        if source_ref == 0:
+            message["arguments"].get("source", {}).get("sourceReference", 0)
```
```suggestion
            source_ref = message["arguments"].get("source", {}).get("sourceReference", 0)
```

vidartf replied:
@r3m0t Thanks for testing. This change is meant to solve this for non-vscode usages as well 👍

Carreau commented:
it's been 4+ years; closing
